### PR TITLE
chore: Allow Babel 8 types

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Ref: https://github.com/babel/babel/pull/15843

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15880"></a>

### CODE REVIEW

**Comment by liuxingbaoyu** [comment]
CI errors are good.

File: packages/babel-generator/src/generators/expressions.ts [hunk 1]
```diff
@@ -81,21 +81,29 @@ export function NewExpression(
...
   this.print(node.typeParameters, node); // TS
 
-  if (node.optional) {
-    // TODO: This can never happen
-    this.token("?.");
+  if (!process.env.BABEL_8_BREAKING) {
+    // @ts-ignore(Babel 7 vs Babel 8) Babel 7 AST
+    if (node.optional) {
+      // TODO: This can never happen
+      this.token("?.");
```
Comment (nicolo-ribaudo) [inline packages/babel-generator/src/generators/expressions.ts:103]: If this can never happen, can we just delete this case? `new a?.()` is invalid syntax.

**Review by nicolo-ribaudo — COMMENTED** [review]
At some point we should also check if the generated types are good, and ideally release them in Babel 8.

File: packages/babel-generator/src/printer.ts [hunk 2]
```diff
@@ -640,16 +640,15 @@ class Printer {
...
-      ];
+    const printMethod = this[
+      nodeType as Exclude<
+        t.Node["type"],
+        // removed
+        | "Noop"
+        // renamed
+        | t.DeprecatedAliases["type"]
+      >
+    ] as (this: Printer, node: t.Node, parent?: t.Node) => void;
```
Comment (armano2) [inline packages/babel-generator/src/printer.ts:651]:
```suggestion
    ] as undefined | (this: Printer, node: t.Node, parent?: t.Node) => void;
```

File: packages/babel-generator/src/generators/expressions.ts (hunk 1 above)
Comment (liuxingbaoyu) [reply packages/babel-generator/src/generators/expressions.ts:103]: Sorry I wasn't clear on this...
But since you said it, I'll remove it.

**Comment by liuxingbaoyu** [comment]
> At some point we should also check if the generated types are good, and ideally release them in Babel 8.

Do you mean we should review babel8's type changes?
I agree with that, maybe we can add a CI to automatically generate babel8 source code snapshots.
If you and @JLHwung think it's good, I'll try it.

**Comment by liuxingbaoyu** [comment]
It seems that there is no `secrets.BOT_TOKEN` in the `CI` action, is there any reason I don't know? Thanks!

**Comment by liuxingbaoyu** [comment]
Expected failure.
https://github.com/babel/babel/actions/runs/6320627582/job/17163430584?pr=15880

**Comment by liuxingbaoyu** [comment]
This PR is ready for review, now we can use `make update-source-snapshots` to update the babel 8 source snapshots.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -262,6 +262,8 @@ jobs:
 
   build-babel8:
     name: Build Babel 8 Artifacts
+    permissions:
+      contents: write
```
Comment (liuxingbaoyu) [inline .github/workflows/ci.yml:266]:
```suggestion
```

File: scripts/integration-tests/e2e-prettier.sh [hunk 4]
```diff
@@ -43,6 +35,15 @@ startLocalRegistry "$root"/verdaccio-config.yml
 yarn install --no-immutable
 yarn info
 
+if [ "$BABEL_8_BREAKING" = true ] ; then
+  # Based on https://github.com/prettier/prettier/pull/15157
+  sed -i 's/const getChalk = () => chalk/default (code) => code/' scripts/build/shims/babel-highlight.js
+  sed -i 's/const generate = babelGenerator.default/const generate = babelGenerator/' scripts/build/transform/index.js
+  sed -i 's/,"updateContext":null//g' tests/integration/__tests__/__snapshots__/debug-print-ast.js.snap
+  rm tests/unit/__snapshots__/visitor-keys.js.snap
+  sed -i 's/node.type !== "MemberExpression" ||/node.type !== "MemberExpression" || \/\/@ts-expect-error/g' src/language-js/utils/is-node-matches.js
```
Comment (liuxingbaoyu) [inline scripts/integration-tests/e2e-prettier.sh:44]: image

**Comment by liuxingbaoyu** [comment]
This PR is too old, I will open a new PR if necessary.

**Pull request closed without merging**